services/wfm/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/delete/actions/CompleteFlowPathRemovalAction.java

```diff
@@ -0,0 +1,134 @@
+import org.openkilda.model.FlowPath;
+import org.openkilda.model.PathId;
+import org.openkilda.model.SwitchId;
+import org.openkilda.persistence.PersistenceManager;
+import org.openkilda.persistence.RecoverablePersistenceException;
+import org.openkilda.persistence.repositories.IslRepository;
+import org.openkilda.wfm.share.history.model.FlowHistoryData;
+import org.openkilda.wfm.share.history.model.FlowHistoryHolder;
+import org.openkilda.wfm.topology.flowhs.fsm.common.action.FlowProcessingAction;
+import org.openkilda.wfm.topology.flowhs.fsm.delete.FlowDeleteContext;
+import org.openkilda.wfm.topology.flowhs.fsm.delete.FlowDeleteFsm;
+import org.openkilda.wfm.topology.flowhs.fsm.delete.FlowDeleteFsm.Event;
+import org.openkilda.wfm.topology.flowhs.fsm.delete.FlowDeleteFsm.State;
+
+import lombok.extern.slf4j.Slf4j;
+import net.jodah.failsafe.RetryPolicy;
+import org.neo4j.driver.v1.exceptions.ClientException;
+
+import java.time.Instant;
+import java.util.Objects;
+import java.util.Optional;
+
+@Slf4j
+public class CompleteFlowPathRemovalAction extends
+        FlowProcessingAction<FlowDeleteFsm, State, Event, FlowDeleteContext> {
+    private final int transactionRetriesLimit;
+
+    private final IslRepository islRepository;
+
+    public CompleteFlowPathRemovalAction(PersistenceManager persistenceManager, int transactionRetriesLimit) {
+        super(persistenceManager);
+        islRepository = persistenceManager.getRepositoryFactory().createIslRepository();
+        this.transactionRetriesLimit = transactionRetriesLimit;
+    }
+
+    @Override
+    protected void perform(State from, State to, Event event, FlowDeleteContext context, FlowDeleteFsm stateMachine) {
+        RetryPolicy retryPolicy = new RetryPolicy()
+                .retryOn(RecoverablePersistenceException.class)
+                .retryOn(ClientException.class)
+                .withMaxRetries(transactionRetriesLimit);
+
+        persistenceManager.getTransactionManager().doInTransaction(retryPolicy, () -> {
+            Flow flow = getFlow(stateMachine.getFlowId());
+            FlowPath[] paths = flow.getPaths().stream().filter(Objects::nonNull).toArray(FlowPath[]::new);
+
+            flowPathRepository.lockInvolvedSwitches(paths);
+
+            for (FlowPath path : paths) {
+                if (path.isForward()) {
+                    // Try to remove as paired first
+                    try {
+                        PathId oppositePathId = flow.getOppositePathId(path.getPathId());
+                        if (oppositePathId != null) {
+                            Optional<FlowPath> oppositePath = flow.getPath(oppositePathId);
+                            if (oppositePath.isPresent()) {
+                                log.debug("Removing the flow paths {} / {}", path.getPathId(), oppositePathId);
+                                flowPathRepository.delete(path);
+                                flowPathRepository.delete(oppositePath.get());
```

Comment:
Do we need to `updateIslsForFlowPath` here?

Reply:
Nice catch! Fixed, although it has been refactored in h-n-s-update.